File: dotnet/fs/example/Samples/STMTest.fs
// File: STMTest.fs
// Project: Samples
// Created Date: 29/08/2023
// Author: Shun Suzuki
// -----
// Last Modified: 25/10/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
// 

namespace Samples

open System
open AUTD3Sharp
open AUTD3Sharp.STM
open AUTD3Sharp.Gain
open AUTD3Sharp.Modulation
open AUTD3Sharp.Utils
open System.Threading.Tasks

module STMTest =
    let GainSTMTest (autd : Controller) = 
        (Silencer.Disable()) |> autd.Send |> ignore;

        (new Static()) |> autd.Send |> ignore;
        
        let center = autd.Geometry.Center + Vector3d(0, 0, 150);
        let stm = 
            [0..199]
            |> List.map (fun i -> (2.0 * AUTD3.Pi * (float)i / 200.0))
            |> List.map (fun theta -> (center + 30.0 * Vector3d(cos(theta), sin(theta), 0.0)))
            |> List.map (fun p -> (new Focus(p)))
            |> List.fold (fun (acc: GainSTM) v -> acc.AddGain v) (new GainSTM(1.))

        printfn $"Actual frequency is {stm.Frequency}";
        (stm )|> autd.Send |> ignore

    let FocusSTMTest (autd : Controller) = 
        (Silencer.Disable()) |> autd.Send |> ignore;

        (new Static()) |> autd.Send |> ignore;
        
        let center = autd.Geometry.Center + Vector3d(0, 0, 150);
        let stm = 
            [0..199]
            |> List.map (fun i -> (2.0 * AUTD3.Pi * (float)i / 200.0))
            |> List.map (fun theta -> (center + 30.0 * Vector3d(cos(theta), sin(theta), 0.0)))
            |> List.fold (fun (acc: FocusSTM) v -> acc.AddFocus v) (new FocusSTM(1.))

        printfn $"Actual frequency is {stm.Frequency}";
        (stm)|> autd.Send |> ignore

File: dotnet/cs/example/Samples/STMTest.cs
/*
 * File: STMTest.cs
 * Project: Samples
 * Created Date: 29/08/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 15/09/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

using AUTD3Sharp;
using AUTD3Sharp.Gain;
using AUTD3Sharp.STM;
using AUTD3Sharp.Modulation;
using AUTD3Sharp.Utils;

namespace Samples;

internal static class GainSTMTest
{
    public static void Test(Controller autd)
    {
        var config = Silencer.Disable();
        autd.Send(config);

        var m = new Static();

        var center = autd.Geometry.Center + new Vector3d(0, 0, 150);
        const int pointNum = 50;
        const double radius = 30.0;
        var stm = new GainSTM(1.0).AddGainsFromIter(Enumerable.Range(0, pointNum).Select(i =>
        {
            var theta = 2.0 * Math.PI * i / pointNum;
            return new Focus(center + radius * new Vector3d(Math.Cos(theta), Math.Sin(theta), 0));
        }));

        Console.WriteLine($"Actual frequency is {stm.Frequency}");
        autd.Send((m, stm));
    }
}

internal static class FocusSTMTest
{
    public static void Test(Controller autd)
    {
        var config = Silencer.Disable();
        autd.Send(config);

        var mod = new Static();
        autd.Send(mod);

        var center = autd.Geometry.Center + new Vector3d(0, 0, 150);
        const int pointNum = 200;
        const double radius = 30.0;
        var stm = new FocusSTM(1).AddFociFromIter(Enumerable.Range(0, pointNum).Select(i =>
        {
            var theta = 2.0 * Math.PI * i / pointNum;
            return center + radius * new Vector3d(Math.Cos(theta), Math.Sin(theta), 0);
        }));

        Console.WriteLine($"Actual frequency is {stm.Frequency}");
        autd.Send(stm);
    }
}

internal static class SoftwareSTMTest
{
    public static void Test(Controller autd)
    {
        var config = Silencer.Disable();
        autd.Send(config);

        var mod = new Static();
        autd.Send(mod);

        var fin = false;
        var th = Task.Run(() =>
        {
            Console.WriteLine("press enter to stop software stm...");
            Console.ReadKey(true);

            fin = true;
        });


        var center = autd.Geometry.Center + new Vector3d(0, 0, 150);
        const double freq = 1.0;
        const int pointNum = 200;
        const double radius = 30.0;
        autd.SoftwareSTM((cnt, i, _) =>
        {
            if (fin) return false;

            var theta = 2.0 * Math.PI * i / pointNum;
            var p = center + radius * new Vector3d(Math.Cos(theta), Math.Sin(theta), 0);
            try
            {
                return cnt.Send(new Focus(center + p));
            }
            catch (Exception)
            {
                return false;
            }
        }).WithTimerStrategy(TimerStrategy.NativeTimer).Start(TimeSpan.FromSeconds(1.0 / freq / pointNum));

        th.Wait();
    }
}
